h2/src/main/org/h2/mvstore/db/MVSecondaryIndex.java

```diff
@@ -77,84 +76,91 @@ public MVSecondaryIndex(Database db, MVTable table, int id, String indexName,
+
             @Override
-            public int compareTo(Source o) {
-                int comp = value.compareTo(o.value, compareMode);
-                if (comp == 0) {
-                    comp = sourceId - o.sourceId;
-                }
-                return comp;
+            public int compare(Source one, Source two) {
+                return one.currentRowData.compareTo(two.currentRowData, compareMode);
             }
         }
-        TreeSet<Source> sources = new TreeSet<>();
-        for (int i = 0; i < bufferNames.size(); i++) {
-            MVMap<Value, Value> map = openMap(bufferNames.get(i));
-            Iterator<Value> it = map.keyIterator(null);
-            if (it.hasNext()) {
-                Source s = new Source();
-                s.value = it.next();
-                s.next = it;
-                s.sourceId = i;
-                sources.add(s);
+    }
+
+    @Override
+    public void addBufferedRows(List<String> bufferNames) {
+        ArrayList<String> mapNames = New.arrayList(bufferNames);
+        CompareMode compareMode = database.getCompareMode();
+        int buffersCount = bufferNames.size();
+        Queue<Source> queue = new PriorityQueue<>(buffersCount, new Source.Comparator(compareMode));
+        for (String bufferName : bufferNames) {
+            Iterator<ValueArray> iter = openMap(bufferName).keyIterator(null);
+            if (iter.hasNext()) {
+                queue.add(new Source(iter));
             }
         }
+
         try {
-            while (true) {
-                Source s = sources.first();
-                Value v = s.value;
+            while (!queue.isEmpty()) {
+                Source s = queue.remove();
+                ValueArray rowData = s.next();
 
                 if (indexType.isUnique()) {
-                    Value[] array = ((ValueArray) v).getList();
+                    Value[] array = rowData.getList();
                     // don't change the original value
                     array = array.clone();
                     array[keyColumns - 1] = ValueLong.get(Long.MIN_VALUE);
                     ValueArray unique = ValueArray.get(array);
-                    SearchRow row = convertToSearchRow((ValueArray) v);
+                    SearchRow row = convertToSearchRow(rowData);
                     checkUnique(row, dataMap, unique);
                 }
 
-                dataMap.putCommitted(v, ValueNull.INSTANCE);
+                dataMap.putCommitted(rowData, ValueNull.INSTANCE);
 
-                Iterator<Value> it = s.next;
-                if (!it.hasNext()) {
-                    sources.remove(s);
-                    if (sources.size() == 0) {
-                        break;
-                    }
-                } else {
-                    Value nextValue = it.next();
-                    sources.remove(s);
-                    s.value = nextValue;
-                    sources.add(s);
+                if (s.hasNext()) {
+                    queue.offer(s);
```

Comment:
seems like we're doing a lot of work removing the first element and then just putting it back again?

Reply:
Yep, thats how it works. It seems like noop, but every time you put that iterator (Source) back in, it falls into a different place, since current row had advanced, and comparison is based on it. I haven't change that aspect actually, it was like that before, just with tree instead of heap.